### PR TITLE
[FIX JENKINS-33411] Use scrolling tabs

### PR DESCRIPTION
**Pull request by gusreiber**
**Scrolling tabs**
https://groups.google.com/forum/#!topic/jenkinsci-dev/g6lOrMFeO3Q
https://t.co/dLFRwXiDVI

@jenkinsci/code-reviewers 

https://issues.jenkins-ci.org/browse/JENKINS-33411
Because today's Jenkins config uses scrolling and some users have expressed an interest in preserving that behavior, this change modifies the tabs to automatically scroll the page to header anchor points, rather than hiding the content of the page as is the case with the traditional tab toggle.
This same behavior will be applied to the initial create items page as part of a separate issue.

### CODE REVIEW

**Comment by daniel-beck** [comment]
Do you have a reference for users who prefer the existing behavior?

Also, wouldn't it be possible to toggle between tabbed scrolling and tabbed tabs? Today we're toggling between classic scrolling and tabbed tabs, after all.

**Comment by swashbuck1r** [comment]
Gus's video rambles on about improvements to the New Item stuff, which isn't directly related to this PR.  The part of the video that is related to this PR is at 4:29 (https://youtu.be/LEKMMZYg9B4?t=269).

I think this is a nice improvement to the new tabs UX for the following reasons:

* The recently added tabs in Jenkins 2 have hidden config fields so you need to click around to figure out what fields need to be set.  As a newbie Jenkins user, this hides some of the upfront complexity, but it also makes it harder for me to find/discover build options since I can no longer scroll to browse them. Note: the search options doesn't help a newbie since I can't search fields that I don't know the names of yet.
* This PR preserves the long-form scrolling behavior of the classic Jenkins config page, while also interweaving the new the tabbed-behavior of recent @tfennelly changes for Jenkins-2.  The tabs become more like a TOC that jumps you to the relevant section, but still keeps the nice look of the tabs. IMO, this is the best of both worlds (long form and tabs).
* This hybrid tabs&long-form approach removes the need for the "switch back to the old way" toggle that was needed in the existing approach to tabs.
* The long-form approach is better for mobile devices, but I don't think managing the config page is really common on a mobile device, so this isn't a strong value add.


Concern:  This makes the tabs JS more complicated, so we need some real testing of this to make sure it doesn't mess things up. Does this need more tests?

**Comment by swashbuck1r** [comment]
@daniel-beck toggles to make it behave different ways adds complexity and increases surface areas that  may break.  IMO, we should commit to one navigation behavior on the config page and make it work great rather than supporting 4 different ways to navigate on the config page.

**Comment by gusreiber** [comment]
@daniel-beck I think toggles atop toggles is a bad idea, in general. It adds more options/complexity for the user and forces us to essentially write 2 different versions of the same feature. The benefit of the scrolling tabs is that they are both scrolling and tabs.

**Comment by swashbuck1r** [comment]
+1 to Gus - this PR adds a way for the config page to work more like classic Jenkins, while also adding the niceties of tabbing, so a "revert to old behavior is no longer needed.   Adding a switch to toggle between different tab modes is crazy madness :)

**Comment by gusreiber** [comment]
@daniel-beck, @ari, Ari Maniatis, is the most recent community member to mention some reservations about classic tabs. I think his opinion on the subject is particularly valuable, because he has actually issued PRs specifically around this issue of navigating the config pages: https://github.com/jenkinsci/jenkins/pull/1746#issuecomment-114029138

I find his final comment arguing against both accordions and tabs very compelling. If you are at the keyboard, it can be really annoying to switch to the mouse, easier to just tab through. Also, the browser find works with the scrolling, but not tabs or accordions.

Ari isn't the first in my experience to voice these sorts of concerns about tabs, just the most recent and most articulate. 

...if I need to, I can dig up more comments back Ari's view. Likely from our own sometimes lengthy discussion threads on the Google groups.

**Comment by gusreiber** [comment]
branch will reopen when I fix the existing tests and add some new ones.

**Comment by gusreiber** [comment]
@review-by-bees

**Comment by gusreiber** [comment]
@reviewbybees

**Comment by daniel-beck** [comment]
Seems wrong to always display the next section's title. With long build sections, this can lead to the selected tab being "Post-build actions" despite seeing only build steps.

> screen shot

Also, Advanced Project Options wasn't its own tab-tab (weirdly short), but should be it's own scroll-tab.

**Comment by daniel-beck** [comment]
Same issue with the tabs sometimes not appearing as PR 2096: [Video](https://dl.dropboxusercontent.com/u/29853/PR2106.mov)

**Comment by ari** [comment]
@gusreiber and others: since I've been mentioned here, I might have one more comment to make. 

* make the core design as absolutely minimalist as possible. Stop designing stuff... you are great developers, but rubbish graphic designers. Embrace that!
* make it easier for others to skin Jenkins. Mostly this is about replacing a lot of inline styles and unhelpful html elements with proper css styles. Remove <b> tags. "style" attributes should be removed everywhere (they are particularly used around images). Move images into css. Sementic markup instead of hardcoded styling. Making your core skin simpler (with less css) will also make it easier for others to over-ride and extend.
* Bring the concepts of a plugin like https://wiki.jenkins-ci.org/display/JENKINS/Simple+Theme+Plugin into the core of Jenkins. Make it easy to bundle js with that as well (to drive accordions and other UI elements)
* Promote the plugins available and make it easier for people to use them. For example, my favourite https://jenkins-contrib-themes.github.io/jenkins-neo-theme/ That project happily accepted my PR and I now use it for our system.

Then maybe a few years from now users will have several dozen nice skins to choose from. Maybe the Jenkins project will adopt one of them as a new default.

I hope this comment is taken in the way it is intended: as a way to cut through the bikeshed and reduce the conflicting opinions by making it easier for others to implement their vision and allowing them to succeed or fail.

**Comment by daniel-beck** [comment]
Tested the PR as of 7f5bfe1, and it does not look like the last section ever gets selected when scrolling. Here I am in a very large *Post-Build Actions* section (email-ext will do that :wink:), and the selected tab is still *Build*, and clicking the *Post-Build Actions* tab will scroll up quite a bit.

screen shot

File: core/src/main/resources/lib/layout/layout.jelly [hunk 1]
```diff
@@ -37,6 +37,9 @@ THE SOFTWARE.
       If non-null and not "false", auto refresh is disabled on this page.
       This is necessary for pages that include forms.
     </st:attribute>
+    <st:attribute name="cssclass" deprecated="true">
```
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/layout.jelly]: You're deprecating something right from the start? Really?

**Comment by gusreiber** [comment]
@daniel-beck when the final sections are short, the last tab cannot be selected by scrolling. This is the correct behavior. If you click the last item, the tab is highlighted.

**Comment by kzantow** [comment]
@daniel-beck :+1: @gusreiber, this is the "expected" behavior of scrollspy

**Comment by gusreiber** [comment]
@kzantow @daniel-beck @swashbuck1r  I can check for the bottom of the page: http://stackoverflow.com/questions/3898130/check-if-a-user-has-scrolled-to-the-bottom ...and then set the last tab. But this will then require some additional knock-on hackery as I pick an arbitrary spot at which I turn off the last tab when the user scrolls back up off the bottom. I can't really argue which behavior is most 'correct', but I can say that we will be piling on extra code to get a behavior that isn't obviously better. Thus I would wait until we get feedback from real users. 

I could inject extra white-space at the bottom of the page, such that the section was always big enough to trigger the tab, but then the submit button is farther away from the form it controls without a clear reason for the space separation. I think that is also worse than leaving it how it is.

**Comment by daniel-beck** [comment]
@gusreiber Except the last section is gigantic. See the screenshot. All you see is a single post-build action, and yet "Build" is selected. I'm at the bottom of the page.

**Comment by kzantow** [comment]
Sorry, I was not clear, @gusreiber @daniel-beck : the expected behavior of scrollspy is that the section at the top of the visible area is 'highlighted'. If there are sections smaller, they won't trigger the 'highlight' to change on the tabs. But it looks like this case actually does have an issue, as the post-build actions section is at the start of the visible area. Is there anything pushing the first section down the scrollable area? Sometime this messes up the scrollspy behavior.

**Comment by daniel-beck** [comment]
FWIW It's only the last section that behaves like this in my testing. The others are currently fine.

**Comment by gusreiber** [comment]
taking a look.

**Comment by daniel-beck** [comment]
Looks good. I suggest to get rid of the previous tab selection in a followup PR. Otherwise this seems to work quite nicely.

File: core/src/main/resources/lib/form/select/select.js [hunk 2]
```diff
@@ -43,6 +43,9 @@ function updateListBox(listBox,url,config) {
 Behaviour.specify("SELECT.select", 'select', 1000, function(e) {
 
         function hasChanged(selectEl, originalValue) {
+            // seems like a race condition allows this to fire before the 'selectEl' is defined. If that happens, exit..
+            if(!selectEl || !selectEl.options || !selectEl[0])
```
Comment (kzantow) [inline core/src/main/resources/lib/form/select/select.js]: assume this should be:
if(!selectEl || !selectEl.options || !selectEl.options[0])

File: war/src/main/js/widgets/config/model/ConfigTableMetaData.js [hunk 3]
```diff
@@ -128,6 +128,16 @@ ConfigTableMetaData.prototype.addFindWidget = function() {
 
     var findTimeout;
     thisTMD.findInput.keydown(function() {
+      
+      //keydown needs to wait a pulse for the value to be applied to the input.
+      setTimeout(function(){
```
Comment (kzantow) [inline war/src/main/js/widgets/config/model/ConfigTableMetaData.js]: yes, ick. any reason not to do keyup?

File: war/src/main/js/widgets/config/model/ConfigTableMetaData.js [hunk 4]
```diff
@@ -255,13 +278,14 @@ ConfigTableMetaData.prototype.showSection = function(section) {
...
-ConfigTableMetaData.prototype.deactivateActiveSection = function() {
+ConfigTableMetaData.prototype.deactivateActiveSection = function(hideRows) {
     var topRows = this.getTopRows();
     var $ = jQD.getJQuery();
 
     $('.config-section-activator.active', this.activatorContainer).removeClass('active');
     topRows.filter('.active').removeClass('active');
-    topRows.hide();
+    if(hideRows) 
+    	{topRows.hide();}
```
Comment (kzantow) [inline war/src/main/js/widgets/config/model/ConfigTableMetaData.js]: this syntax... lint police! :)

**Comment by gusreiber** [comment]
Made fixes based on @kzantow 's comments.

File: core/src/main/resources/lib/form/select/select.js [hunk 5]
```diff
@@ -44,7 +44,11 @@ Behaviour.specify("SELECT.select", 'select', 1000, function(e) {
 
         function hasChanged(selectEl, originalValue) {
             // seems like a race condition allows this to fire before the 'selectEl' is defined. If that happens, exit..
+<<<<<<< HEAD
```
Comment (daniel-beck) [inline core/src/main/resources/lib/form/select/select.js]: Unfinished merge.

**Comment by jglick** [comment]
Why is the configuration screen width restricted to ⅓ of my browser’s visible display area? I did not buy a high-resolution display just to waste 70% on blank white space. :-1:
screenshot

**Comment by jglick** [comment]
The 2.0 branch does not have this problem, so it is a regression in this PR.

**Comment by gusreiber** [comment]
@jglick You are running a 4k display and large fonts, yeah? Can you give me some of those specifics so I can repro?

**Comment by jglick** [comment]
@gusreiber 3k display @ 2880×1620, Chrome, 20pt minimum font size. Anyway I think that is beside the point: the page should use the available screen width, whatever that is.

File: core/src/main/resources/lib/layout/layout.jelly [hunk 6]
```diff
@@ -96,7 +99,13 @@ ${h.initPageVariables(context)}
     </j:if>
     <link rel="shortcut icon" href="${resURL}/favicon.ico" type="image/vnd.microsoft.icon" />
     <link rel="mask-icon" href="${rootURL}/images/mask-icon.svg" color="black" />
-
+    
+    <!-- should the sidebar be hidden on load? -->
+    <style>
+      body.j-hide-left #side-panel{width:0; padding:0; overflow:hidden;}
+      body.j-hide-left #main-panel{margin:0 auto; max-width:900px;}
```
Comment (jglick) [inline core/src/main/resources/lib/layout/layout.jelly]: Try just deleting `max-width`.

File: war/src/main/js/widgets/config/tabbar.less [hunk 7]
```diff
@@ -1,6 +1,93 @@
 /*
  * Tab bar specific rules.
  */
+body.add-item.hide-side  #side-panel{width:0; padding:0; overflow:hidden; transition: all 0.5s ease;}
+body.add-item.hide-side  #main-panel{margin:0 auto; max-width:900px; transition: all 0.5s ease;}
```
Comment (jglick) [inline war/src/main/js/widgets/config/tabbar.less]: ditto

**Comment by gusreiber** [comment]
@jglick ....yes. you think that. 20pt minimum font size is ridiculously huge. But I can likely get it to work. There is a readable width for text, so no, it shouldn't go all the way across your giant screen.

**Comment by daniel-beck** [comment]
@jglick @gusreiber @kohsuke Anyone opposed to doing the screen width changes in a follow-up PR? Would be nice to get this out to people for testing and feedback…

**Comment by jglick** [comment]
> 20pt minimum font size is ridiculously huge.

Not actually. The pixels are small.

> There is a readable width for text

For newspapers perhaps. For novels. When I am configuring, say,

    <goals>clean install -e -U -Dignore.random.failures ${findbugs_run == null || not findbugs_run ? '' : '-Dfindbugs.failOnError=false -Dmaven.findbugs.failure.strict=false findbugs:findbugs'} ${concurrency == null || concurrency == '' ? '' : '-Dconcurrency=' + concurrency}  ${skip_javadoc == null || not skip_javadoc ? '' : '-Dmaven.javadoc.skip=true'} </goals>

I want to use my screen. All of it.

> Anyone opposed to doing the screen width changes in a follow-up PR?

Personally I find the improvement this makes to pale in significance next to the loss of space. You cannot convince me to move into a Manhattan studio by arguing that the décor is nice.

**Comment by gusreiber** [comment]
@jglick @daniel-beck @kohsuke ...I am working on a fix. PS. A lot of people do live in Manhattan and other cities quite happily.

**Comment by kohsuke** [comment]
> You cannot convince me to move into a Manhattan studio by arguing that the décor is nice.

Speaks like a true Bostonian!

**Comment by kzantow** [comment]
I have to disagree about the loss of space being a significant issue when you take into consideration the tabs otherwise are hiding a significant amount of content from the user. I'd definitely agree with Daniel's suggestion and :+1: getting this merged, getting feedback, and adjusting accordingly in a follow-on PR before the official release.

**Comment by jglick** [comment]
> the tabs otherwise are hiding a significant amount of content from the user

Potentially, but only _vertically_, and if you

**Comment by kohsuke** [comment]
Let's get this in now in time for alpha-3 and I'm sure @gusreiber will be happy to work on any follow up issues.

The point of getting these features into 2.0 previews is so that people can use them and find problems. It's better to get stuff in the hands of people early so that we have time to work on fixes.

**Comment by rtyler** [comment]
I'm in agreement with @kohsuke, for an alpha release I think the horizontal space issue is _okay_ but I also agree with @jglick that it's not trivial and we should address it prior to the beta.

**Comment by gusreiber** [comment]
...so I feel pretty strongly that no maximum width here is no the correct way to go. That having been said, Jesse's screen looked broken and he should definitely be getting more screen than he is. So I have changed the max width to be relative to font-size, 75em, to be exact. That makes the input and select boxes wider than I think is comfortable to look at, but hopefully tolerable for Jesse.

**Comment by jglick** [comment]
Still broken with your latest changes.

still-broken

With

```diff
diff --git a/core/src/main/resources/lib/layout/layout.jelly b/core/src/main/resources/lib/layout/layout.jelly
index 9375552..9397ed8 100644
--- a/core/src/main/resources/lib/layout/layout.jelly
+++ b/core/src/main/resources/lib/layout/layout.jelly
@@ -103,7 +103,7 @@ ${h.initPageVariables(context)}
     <!-- should the sidebar be hidden on load? -->
     <style>
       body.j-hide-left #side-panel{width:0; padding:0; overflow:hidden;}
-      body.j-hide-left #main-panel{margin:0 auto; max-width:75em;}
+      body.j-hide-left #main-panel{margin:0 auto}
     </style>
     
     <!-- are we running as an unit test? -->
diff --git a/war/src/main/js/widgets/config/tabbar.less b/war/src/main/js/widgets/config/tabbar.less
index 05cbd35..719a0b9 100644
--- a/war/src/main/js/widgets/config/tabbar.less
+++ b/war/src/main/js/widgets/config/tabbar.less
@@ -2,7 +2,7 @@
  * Tab bar specific rules.
  */
 body.add-item.hide-side  #side-panel{width:0; padding:0; overflow:hidden; transition: all 0.5s ease;}
-body.add-item.hide-side  #main-panel{margin:0 auto; max-width:75em; transition: all 0.5s ease;}
+body.add-item.hide-side  #main-panel{margin:0 auto; transition: all 0.5s ease;}
 body.add-item .jenkins-config table tr {display:none;}
 body.add-item .jenkins-config table tr:last-of-type {display:table-row}
 body.add-item .jenkins-config-widgets .form-config.tabBarFrame .tabBar .tab.active {background:@bright; border-bottom-color:@bright}
```

it looks OK to me:

no-max-width

By comparison, unpatched 2.0



and 1.x

1 x

I did find a bug even in 1.x: if you have a very long line in the shell script textarea, rather than scrollbars appearing inside that textarea (or it simply line-wrapping at word boundaries like I would like), the whole form stretches out past the screen boundary and gets a page-scope scrollbar.

**Comment by kohsuke** [comment]
I captured what Jesse is describing as [JENKINS-33471](https://issues.jenkins-ci.org/browse/JENKINS-33471)

**Comment by kohsuke** [comment]
This change is in 2.0-alpha-3

**Comment by tfennelly** [comment]
@gusreiber @swashbuck1r @daniel-beck @kohsuke Sorry guys ... I missed this, but I have to say that, aside from the styling improvements (which I do think are great), I really do not find the scrollspy to be all that much of an improvement over no nav. I find the argument that it's a better first time UX (because new users can see everything that they might want to configure) doesn't add up with what has long been recognised as being one of the first time UX issues with config pages i.e. that new users are overwhelmed with the toilet roll config (with being able to see everything they might want to config).

There's more that could be said but I think it's probably a waste of time saying it.

:-1: for adding zero tests.

**Comment by daniel-beck** [comment]
The config form width issue has been addressed towards beta 1.

**Pull request merged**